### ISSUE TITLE
the first write on a TcpStream after connection has been lost does not returns an error

### ISSUE THREAD

**Issue by silvadanilo** [tokio-rs/tokio-core#192]
I would expect that a write on a TcpStream must returns an error after that the connection has been lost.
It seems that the error is returned, but only after the second attempt (after the connection is gone) while I would expect the error on the first attempt.
The behaviour is the same using `.send()` on a `SplitSink` obtained with the `.framed()` method or using a `write_all` method on a `WriteHalf` obtained with a `.split()` method.

I posted a repository to reproduce the possible bug
https://github.com/silvadanilo/rust-testing-sending-error/blob/master/src/main.rs
It tries to send a message to a remote server every 2 seconds.
To run it a server should be instantiate before running (on linux it can be instatiated with nc: `nc -k -l 9876`

I also posted a [video](https://www.youtube.com/watch?v=P2MLzasaeAE&feature=youtu.be) to show the behaviour.
On this video my expected behaviuour would be an error after sending `Message 3`, instead it has occured after sending `Message 4`.

**Comment by alexcrichton** [comment]
Out of curiosity, are you on Windows as well? 

If so I believe I know where this is happening internally, but just want to confirm :).

Also if you can get the most up-to-date version of mio and run `RUST_LOG=mio=trace` when the behavior reproduces that'd also be mega-helpful (it's a massive amount of output, but should have nuggets of info)

**Comment by silvadanilo** [comment]
windows?!? what is it?
no no, on linux :)

I had this problem on old version of rust and also in the latest, the version of mio (founded in Cargo.lock) is `version = "0.6.5"`

unfortunately running it with `RUST_LOG=mio=trace` doesnt't add any useful information:

(for this output server has been stopped after `message 1`)
```
TRACE:mio::poll: registering with poller
TRACE:mio::poll: registering with poller
TRACE:mio::poll: registering with poller
TRACE:mio::poll: registering with poller
Interval
before sending `Message -> 1`
sent
Interval
before sending `Message -> 2`
sent
Interval
before sending `Message -> 3`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { repr: Os { code: 32, message: "Broken pipe" } }', /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libcore/result.rs:868
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by silvadanilo** [comment]
it it can be useful: if the future returned is linked with the future created on the reader (through the select method like you do [in the chat example](https://github.com/tokio-rs/tokio-core/blob/master/examples/chat.rs#L122)) it works.
It seems that the `SplitStream`/`ReadHalf` will properly closed.

**Comment by alexcrichton** [comment]
Cool thanks for the clarification. In retrospect I should have looked at your video, then it would have been self explanatory!

In any case I've reproduced and what I witnessed was:

* We're not swallowing up errors in `write` at all, the syscall does indeed return success on the first time after the remote disconnects.
* As soon as the write completes, though, then the file descriptor is ready with `EPOLLERR`.
* The program here isn't listening for any events from the file descriptor, so the next event is an internval message which causes the failed write to become apparent.

So I think the problem here may be in the program and not the event loop? In general there's no way of knowing whether a write actually succeeded or not unless you get a confirmation from the remote end, and it looks like this may be a use case where fundamentally the error in write is just deferred until later perhaps?

**Comment by silvadanilo** [comment]
I don't really know, sorry, I'm not a system programming expert.
What I can say is that I would expect an error even if I'm not reading from the SplitStream/Readhalf because the socket is closed and I would expect that my program knows that he cannot write anymore on that socket.

I don't know EPOLL event at the low level, but I run these further tests:

- when I shutdown the server, the socket of the rust program goes in `CLOSE_WAIT` state (checked with `netstat`), so the program knows that he can not write anymore on it, it only have to close it. 
After the next write it effectively close the socket (maybe he should close it before?) 
- I ran the program with `strace -f` command and what I've discovered is:
    - when the server is shutted down the rust program receive this events:
    ```
    <unfinished ...>
    [pid 25507] <... epoll_wait resumed> {{EPOLLIN|EPOLLOUT|EPOLLRDHUP, {u32=2, u64=2}}}, 1024, -1) =  
    [pid 25507] epoll_wait(3, 
    ```
    (it contains `EPOLLRDHUP` that should be the signal that the socket was closed, maybe you should react on that event? Really I don't know, I'm just guessing)

    - after the next write the events are a little bit different, I had found this:
      `[pid 25507] epoll_wait(3, {{EPOLLIN|EPOLLOUT|EPOLLERR|EPOLLHUP|EPOLLRDHUP, {u32=2, u64=2}}}, 1024, 0) = 1` in the trace block. 
    
     let me paste a sample trace slice:
    
--
**Trace of a WRITE when the socket is ESTABLISHED**
```
 <unfinished ...>
[pid 25536] <... futex resumed> )       = -1 ETIMEDOUT (Connection timed out)
[pid 25536] futex(0x7f1671c1e1e0, FUTEX_WAKE_PRIVATE, 1) = 0
[pid 25536] write(5, "\1", 1)           = 1
[pid 25507] <... epoll_wait resumed> {{EPOLLIN, {u32=4294967295, u64=18446744073709551615}}}, 1024, -1) = 1
[pid 25507] read(4, "\1", 128)          = 1
[pid 25507] read(4, 0x7ffe952a2a60, 128) = -1 EAGAIN (Resource temporarily unavailable)
[pid 25536] futex(0x7f1671c1e214, FUTEX_WAIT_PRIVATE, 7, NULL <unfinished ...>
[pid 25507] write(1, "Interval\n", 9Interval
)   = 9
[pid 25507] futex(0x7f1671c1e214, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f1671c1e210, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1} <unfinished ...>
[pid 25536] <... futex resumed> )       = 0
[pid 25507] <... futex resumed> )       = 1
[pid 25536] futex(0x7f1671c1e1e0, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>
[pid 25507] futex(0x7f1671c1e1e0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 25536] <... futex resumed> )       = -1 EAGAIN (Resource temporarily unavailable)
[pid 25507] <... futex resumed> )       = 0
[pid 25536] futex(0x7f1671c1e1e0, FUTEX_WAKE_PRIVATE, 1) = 0
[pid 25507] epoll_wait(3, {}, 1024, 0)  = 0
[pid 25507] write(1, "before sending `Message -> 2`\n", 30before sending `Message -> 2`
) = 30
[pid 25507] sendto(6, "Message -> 2\n", 13, MSG_NOSIGNAL, NULL, 0) = 13
[pid 25507] write(1, "sent\n", 5sent
)       = 5
[pid 25507] epoll_wait(3, {}, 1024, 0)  = 0
[pid 25507] epoll_wait(3,  <unfinished ...>
[pid 25536] futex(0x7f1671c1e214, FUTEX_WAIT_BITSET_PRIVATE, 9, {229335, 103718167}, ffffffff
```
    
--
**When the socket goes in CLOSE_WAIT**
```
<unfinished ...>
[pid 25507] <... epoll_wait resumed> {{EPOLLIN|EPOLLOUT|EPOLLRDHUP, {u32=2, u64=2}}}, 1024, -1) = 1
[pid 25507] epoll_wait(3, 
```
    
--
**Trace of a WRITE when the socket is closed**
```
 <unfinished ...>
[pid 25536] <... futex resumed> )       = -1 ETIMEDOUT (Connection timed out)
[pid 25536] futex(0x7f1671c1e1e0, FUTEX_WAKE_PRIVATE, 1) = 0
[pid 25536] write(5, "\1", 1)           = 1
[pid 25507] <... epoll_wait resumed> {{EPOLLIN, {u32=4294967295, u64=18446744073709551615}}}, 1024, -1) = 1
[pid 25507] read(4, "\1", 128)          = 1
[pid 25507] read(4, 0x7ffe952a2a60, 128) = -1 EAGAIN (Resource temporarily unavailable)
[pid 25536] futex(0x7f1671c1e214, FUTEX_WAIT_PRIVATE, 11, NULL <unfinished ...>
[pid 25507] write(1, "Interval\n", 9Interval
)   = 9
[pid 25507] futex(0x7f1671c1e214, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f1671c1e210, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
[pid 25536] <... futex resumed> )       = 0
[pid 25507] epoll_wait(3, {}, 1024, 0)  = 0
[pid 25536] futex(0x7f1671c1e1e0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 25507] write(1, "before sending `Message -> 3`\n", 30 <unfinished ...>
before sending `Message -> 3`
[pid 25536] <... futex resumed> )       = 0
[pid 25507] <... write resumed> )       = 30
[pid 25507] sendto(6, "Message -> 3\n", 13, MSG_NOSIGNAL, NULL, 0) = 13
[pid 25507] write(1, "sent\n", 5sent
)       = 5
[pid 25507] epoll_wait(3, {{EPOLLIN|EPOLLOUT|EPOLLERR|EPOLLHUP|EPOLLRDHUP, {u32=2, u64=2}}}, 1024, 0) = 1
[pid 25507] epoll_wait(3,  <unfinished ...>
[pid 25536] futex(0x7f1671c1e214, FUTEX_WAIT_BITSET_PRIVATE, 13, {229340, 103645603}, ffffffff
```
    
--
**Trace of a the second WRITE after the socket is closed**
```
) = -1 ETIMEDOUT (Connection timed out)
[pid 25536] futex(0x7f1671c1e1e0, FUTEX_WAKE_PRIVATE, 1) = 0
[pid 25536] write(5, "\1", 1)           = 1
[pid 25507] <... epoll_wait resumed> {{EPOLLIN, {u32=4294967295, u64=18446744073709551615}}}, 1024, -1) = 1
[pid 25507] read(4, "\1", 128)          = 1
[pid 25507] read(4, 0x7ffe952a2a60, 128) = -1 EAGAIN (Resource temporarily unavailable)
[pid 25507] write(1, "Interval\n", 9Interval
)   = 9
[pid 25536] futex(0x7f1671c1e214, FUTEX_WAIT_PRIVATE, 15, NULL <unfinished ...>
[pid 25507] futex(0x7f1671c1e214, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7f1671c1e210, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1} <unfinished ...>
[pid 25536] <... futex resumed> )       = 0
[pid 25507] <... futex resumed> )       = 1
[pid 25536] futex(0x7f1671c1e1e0, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>
[pid 25507] futex(0x7f1671c1e1e0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 25536] <... futex resumed> )       = -1 EAGAIN (Resource temporarily unavailable)
[pid 25507] <... futex resumed> )       = 0
[pid 25536] futex(0x7f1671c1e1e0, FUTEX_WAKE_PRIVATE, 1) = 0
[pid 25507] epoll_wait(3, {}, 1024, 0)  = 0
[pid 25507] write(1, "before sending `Message -> 4`\n", 30before sending `Message -> 4`
) = 30
[pid 25507] sendto(6, "Message -> 4\n", 13, MSG_NOSIGNAL, NULL, 0) = -1 EPIPE (Broken pipe)
[pid 25507] close(6)                    = 0
```

**Comment by silvadanilo** [comment]
> The program here isn't listening for any events from the file descriptor, so the next event is an internval message which causes the failed write to become apparent.

are you suggesting to read from the `SplitStream`/`ReadHalf` ?
If not, how can I listening for events on the fd?

**Comment by alexcrichton** [comment]
There may be a better way to handle this perhaps, but I'm not really sure what the best practice is. The problem is that the syscall returns success on the first write after the connection is severed. We then immediately *later* learn that the socket is in an erroneous state and we could call `take_error()` to get it out but we have no one to return it to. I believe any future call to `write` would do roughly the same thing and return an error.

So the recourse for us right now is that there has to be some second operation to happen on a TCP stream. I'm not really sure what this use case is modeling (which may help for figuring out what to do) but some options are:

1. Connection severance typically implies EOF on the read side, which can be used as an event if you're waiting for data to be read.
2. If you're writing a large block of data you're likely waiting for it to become writable again in which case you'll hit the error.
3. We could add a feature "wait for an error to appear" although I'm not quite sure what that would look like.
